### PR TITLE
Improving OWSM inference interface

### PR DESCRIPTION
**Pull request by pyf98**
## What?

- This PR improves the interface for OWSM inference.
  - The speech is automatically padded or trimmed to the fixed length which is consistent with training.
  - Some attributes like speech length and time symbols can be retrieved from `preprocessor_conf`. We no longer need to provide them manually.
  - `lang_sym`, `task_sym` and `predict_time` can be passed as additional arguments when calling `__call__`, which will overwrite the default values in `__init__`. This is more convenient to use.
  - Many redundant arguments are removed from `s2t_inference_language`. `BeamSearch` is also removed. Only a single decoder forward step is performed. And it can return an N-best list of language and (normalized) probability.
- It also implements *some* simple rules to suppress timestamps which is similar to [Whisper](https://github.com/openai/whisper/blob/main/whisper/decoding.py#L441). The implementation defines a new `scorer` that assigns certain tokens with a `-inf` score during `search`. With this modification, it can now predict the first timestamp by itself. Previously, the first timestamp has to be manually set, which is usually inaccurate.

### CODE REVIEW

**Comment by sw005320** [comment]
Thanks, @pyf98!
@jctian98, can you review this PR?

**Comment by pyf98** [comment]
Hi, can we considering merging this soon? I will provide example usage in the project page based on this new interface.

As reported previously, we can significantly improve the long-form ASR performance with some rules and heuristics
`TEDLIUM2 WER: 8.5 -> 5.7 (greedy)`

**Comment by sw005320** [comment]
Yes, but currently, CI is broken.
So, we're checking it.
Sorry about it.

**Comment by ftshijt** [comment]
@sw005320 The CI has passed the previous issue. Should be good to go ahead

**Comment by sw005320** [comment]
Cool!

@pyf98, please deal with the CI error
https://github.com/espnet/espnet/actions/runs/7484131332/job/20534760203?pr=5618

**Comment by pyf98** [comment]
Thanks! It seems that some TTS tests failed

**Comment by sw005320** [comment]
@jctian98, this is a reminder.
Can you review this PR?

File: espnet2/bin/s2t_inference.py [hunk 1]
```diff
@@ -44,6 +43,105 @@
 ]
 
 
+class ScoreFilter(BatchScorerInterface, torch.nn.Module):
+    """Filter scores based on pre-defined rules."""
```
[the remark this answers is not in the document]
Comment (pyf98) [reply espnet2/bin/s2t_inference.py]: I added some more comments below

**Comment by jctian98** [comment]
The code is very clear. I'm ok with it.

A simple discussion for future development:
Do you think the `s2t_inference_language.py` can be a special case of `s2t_inference.py` as long as you pass a constant `maxlen` to beam search so that `decoding one step` exactly means `predicting the lang_id`?

**Comment by pyf98** [comment]
Thanks @jctian98 for the comment.

In the previous version of `s2t_inference_language.py`, I simply copied the code of `s2t_inference.py` and then set `max_len` to be 1, which can predict the language token given `<sos>`. This design has some issues:
- It adds many redundant arguments since we directly reuse `s2t_inference.py` which is designed to predict a fully formated sequence. In the new version, I removed those redundant arguments to avoid confusion. Also, it no longer requires the `BeamSearch` object.
- The current `BatchBeamSearch` only supports `batch_size=1`. In the future, we may want to implement batched language detection. This will be much easier in the new version.

I think we can keep this separate design for now.

**Comment by pyf98** [comment]
One more note is that the order of tokens is: `<sop> prompt<sos><lang><task><time>xxx<eos`.

The language token is between `<sos>` and `<task>`. So, we cannot predict `<lang>` and text in a single autoregressive decoding run. Instead, we need to first predict `<lang>` given `<sos>`, and then predict text given `<sos><lang><task>` where `<task>` is known.

**Comment by sw005320** [comment]
Thanks, @pyf98!
